proposals/010-authorizer.md

```diff
@@ -0,0 +1,450 @@
+     * Determines whether the given {@code subject} is allowed to perform the given {@code actions}.
+     * The implementation must ensure that the returned authorization partitions all the given {@code actions}
+     * between {@link Authorization#allowed()} and {@link Authorization#denied()}.
+     * @param subject The subject.
+     * @param actions The actions.
+     * @return The outcome.
+     */
+    CompletionStage<AuthorizationDecisions> authorize(Subject subject, List<Action> actions);
+}
+```
+
+Note:
+* The asynchronous return type is intended to permit implementations using technologies such as [OPA][OPA] or [OpenFGA][OpenFGA], which would otherwise need to block the thread making the decision over the network.
+* Because of the possibility of higher latency authorizer implementations, call sites should try to pass as many `actions` as possible so as to benefit from batching requests.
+* Unlike Apache Kafka we don't currently include in the Authorizer the responsibility to log authorization attempts.
+
+An `Action` describes something the subject may be attempting to do:
+
+```java
+public record Action(
+                     ResourceType<?> resourceType,
+                     String resourceName) { ... }
+```
+
+Where `ResourceType` has this declaration:
+
+```java
+public interface ResourceType<V extends Enum<V> & ResourceType<V>> { ... }
+```
+
+The scary-looking type parameter bound means a `ResourceType` implementation must be an `enum` rather than `class`, `interface`, etc.
+The `enum`'s elements are the operations that a particular resource type supports. 
+For example we would model a Kafka topic like this:
+
+```java
+enum Topic implements ResourceType<Topic> {
+    READ,
+    WRITE,
+    CREATE,
+    // ...
+}
+
+The `java.lang.Class` of a particular ResourceType then serves to identify the type of a resource in the API.
+Thus, the `Action` instance for describing a topic called "foo" would be created like so:
+
+```java
+new Action(Topic.DESCRIBE, "foo")
+```
+
+The `AuthorizationDecisions` type provides access to the decisions about each of the actions in the query to `Authorizer.authorize()`.
+That is, `Authorizer.authorize()` partitions its `actions` argument into the `allowed` and `denied` lists of the returned `AuthorizationDecisions`.
+
+```java
+public enum Decision {
+    ALLOW,
+    DENY;
+}
+
+public record AuthorizationDecisions(
+                            Subject subject,
+                            List<Action> allowed,
+                            List<Action> denied) {
+
+    public Decision decision(Operation<?> operation, String resourceName) { ... }
+    
+    // ...
+}
+```
+
+### The `kroxylicious-authorizer-acl` module
+
+The new `kroxylicious-authorizer-acl` module will provide a implementation of `Authorizer` using an Access Control List (ACL), called `AclAuthorizer`.
+The ACLs will be defined in a file (external to the proxy configuration file) which enumerates _who_ is allowed to do _what_.
```

Comment:
wonder if we should require the filename or content to encode a language version, might give us more options if we want to make breaking changes but retain support for an old grammar for a while.

Reply:
This is currently supported. I think it's a bit ugly, to be honest, and I think backwards compatible rules language evolution is less of a problem when using a proper grammar, than when you're trying to squeeze this kind of thing into YAML. You could draw a comparison to programming languages like Java where the compiler has been source backwards compatible for a while now. You just have to be prepared to be inventive with the grammar. So I almost wonder about taking it out.